### ISSUE TITLE
Exclude selected operations from INT8 quantized model

### ISSUE THREAD

**Issue by adamp87** [PINTO0309/onnx2tf#578]
### Issue Type

Feature Request

### OS

Linux

### onnx2tf version number

1.17.5

### onnx version number

1.15.0

### onnxruntime version number

1.16.3

### onnxsim (onnx_simplifier) version number

0.4.35

### tensorflow version number

2.13.1

### Download URL for ONNX

pip install ultralytics
yolo export model=yolov8n.pt imgsz=640 format=onnx
# yolo export model=yolov8n.pt data=coco128.yaml imgsz=640 format=tflite int8

### Parameter Replacement JSON

```json
{
}
```

### Description

1. Research and Development
2. Accuracy loss due to a few final operations in YOLO8n head
3. Executed tests show not quantizing a few final operations by splitting the detection head can increase accuracy with ~2.4%
4. Many heads in models could benefit from controlling where the quantization in the compute graph should stop
5. https://github.com/ultralytics/ultralytics/pull/7516#issuecomment-1913636777

I am aware of the `PARAM_REPLACEMENT_FILE` option, but haven't found out how to exclude an operation from being quantized.

It would a nice feature to be able to control which operations should be excluded from quantization. Test show that for YOLO8n accuracy can drop ~2.4% because of the last few operations.

Models larger than 192 pixel cannot be compiled fully for EdgeTPU because of the Transpose. Therefore the final operations are executed anyways on the CPU.
![yolov8n_full_integer_quant_edgetpu tflite](https://github.com/PINTO0309/onnx2tf/assets/7806910/c30e33f2-ec54-4150-b39f-238ddb21e1ac)

**Comment by PINTO0309** [comment]
The yolov8 thread is too long to read.

Clearly state here which operations to skip and from which operations. It's too much trouble to find out.

**Comment by adamp87** [comment]
The idea would be to have a general option, maybe in `PARAM_REPLACEMENT_FILE`, so a user could edit each model accordingly. Yolov8 is just an example where it could be beneficial. For the attached yolov8 model visualization, the Concatenation and every OP after it should be excluded.

**Comment by PINTO0309** [comment]
There is more than one `Concat`. I didn't know which `Concat` he was talking about, so I cut it down appropriately. And I still don't understand what you expect me to do.
```
onnx2tf \
-i yolov8_n.onnx \
-onimc /model.22/Sigmoid_output_0 /model.22/Sub_1_output_0 /model.22/Div_output_0
```
![image](https://github.com/PINTO0309/onnx2tf/assets/33194443/799e4cd7-0801-473f-9572-2f3d4af3b7f4)

**Comment by adamp87** [comment]
Hey,

sorry for the confusion. Let me try to explain with a visual example. This is a toy example, the model makes no sense its just for illustration. 

The left is the fully quantized INT8 model, which is the output of the onnx2tf tool. The right model is what I would like to have. I would like to insert a dequantization and keep everything after dequantization in FP32.
![model](https://github.com/PINTO0309/onnx2tf/assets/7806910/cd36cfcd-a981-4147-807d-83124eed30e7)

**Comment by PINTO0309** [comment]
Are you saying that the position of the inverse quantization should be controllable by onnx2tf? That is impossible. Let me know if you know of any of the TensorFlow converter parameters below that can handle your request.

https://github.com/tensorflow/tensorflow/blob/6e6ca51e99c8d46c401ad11982cbf846c3a4071f/tensorflow/lite/python/lite.py#L634

```python
class TFLiteConverterBase:
  """Converter superclass to share functionality between V1 and V2 converters."""

  # Stores the original model type temporarily to transmit the information
  # from the factory class methods to TFLiteConverterBase init function.
  _original_model_type = conversion_metdata_fb.ModelType.NONE

  def __init__(self):
    self.optimizations = set()
    self.representative_dataset = None
    self.target_spec = TargetSpec()
    self.allow_custom_ops = False
    self.experimental_new_converter = True
    self.experimental_new_quantizer = True
    self.experimental_enable_resource_variables = True
    self._experimental_calibrate_only = False
    self._experimental_sparsify_model = False
    self._experimental_disable_per_channel = False
    self._debug_info = None  # contains the stack traces of all the original
    # nodes in the `GraphDef` to the converter.
    self.saved_model_dir = None
    self._saved_model_tags = None
    self._saved_model_version = 0
    self._saved_model_exported_names = []
    self._tflite_metrics = metrics.TFLiteConverterMetrics()
    self._collected_converter_params = {}
    self.unfold_batchmatmul = False
    self.legalize_custom_tensor_list_ops = False
    self._experimental_lower_tensor_list_ops = True
    self._experimental_default_to_single_batch_in_tensor_list_ops = False
    self._experimental_unfold_large_splat_constant = False
    self._experimental_tf_quantization_mode = None
    # If unset, bias:int32 is by default except 16x8 quant.
    # For 16x8 quant, bias:int64 is used to prevent any overflow by default.
    # The accumulator type will be the same as bias type set by
    # full_integer_quantization_bias_type.
    self._experimental_full_integer_quantization_bias_type = None
    # Provides specs for quantization, whether preset or custom.
    self._experimental_quantization_options = None  # Deprecated
    self.experimental_use_stablehlo_quantizer = False
    # Initializes conversion metadata.
    self.exclude_conversion_metadata = False
    self._metadata = conversion_metdata_fb.ConversionMetadataT()
    self._metadata.environment = conversion_metdata_fb.EnvironmentT()
    self._metadata.options = conversion_metdata_fb.ConversionOptionsT()
    self._metadata.environment.tensorflowVersion = versions.__version__
    self._metadata.environment.modelType = self._get_original_model_type()
    self._experimental_enable_dynamic_update_slice = False
    self._experimental_preserve_assert_op = False
    self._experimental_guarantee_all_funcs_one_use = False

    # When the value is true, the MLIR quantantizer triggers dynamic range
    # quantization in MLIR instead of the old quantizer. Used only if
    # experimental_new_quantizer is on.
    self.experimental_new_dynamic_range_quantizer = True
    # Experimental flag to enable low-bit QAT in 8 bit.
    self._experimental_low_bit_qat = False
    # Experimental flag to add all TF ops (including custom TF ops) to the
    # converted model as flex ops.
    self._experimental_allow_all_select_tf_ops = False

    self._experimental_variable_quantization = False
    self._experimental_disable_fuse_mul_and_fc = False
    self._experimental_use_buffer_offset = False
    self._experimental_reduce_type_precision = False
    self._experimental_qdq_conversion_mode = None

    # Debug parameters
    self.ir_dump_dir = None
    self.ir_dump_pass_regex = None
    self.ir_dump_func_regex = None
    self.enable_timing = None
    self.print_ir_before = None
    self.print_ir_after = None
    self.print_ir_module_scope = None
    self.elide_elementsattrs_if_larger = None
```

**Comment by adamp87** [comment]
I dont know if thats possible with TFLite, but would be good. For example OpenVINOs NNCF convert() function do accept a parameter named `ignored_scope`.

Yes, it is a toy example and toy example are meant to be simple. In YOLO8, after the Concat (not the last before the output) everything should be executed in FP32.

**Comment by PINTO0309** [comment]
I finally understand what you are intending.

I will say it again. It is impossible. Submit a feature request issue to TensorFlow.

**Comment by adamp87** [comment]
Oh, all right, sad to hear. Thanks for your help.

**Comment by EpiX-1** [comment]
Hi  @adamp87,

I think you can achieve what you want by converting your onnx model to keras using the `-oh5` option from onnx2tf.
Then you can quantize your keras model with the TensorFlow API using [tf.lite.experimental.QuantizationDebugger()](https://www.tensorflow.org/api_docs/python/tf/lite/experimental/QuantizationDebugger) which let you specify nodes/operators to skip during the quantization process (check [tf.lite.experimental.QuantizationDebugger()](https://www.tensorflow.org/lite/api_docs/python/tf/lite/experimental/QuantizationDebugOptions)).
Hope this helps.

**Comment by adamp87** [comment]
Hi @EpiX-1,

thank you so much for the suggestion. I think this could go the right way. Sadly Im having an issue and not sure how to go further. Are you familiar with onnx2tf or could you give me a hint what could be the problem?

By running the following in Colab:
```
!git clone https://github.com/adamp87/ultralytics.git
%cd /content/ultralytics
!git checkout tflite_accurate
!pip install -e .
!yolo export model=yolov8n.pt data=coco128.yaml format=tflite imgsz=640
```
I get this error during the init of `QuantizationDebugger`: 
`'/model.10/Resize' is not a valid root scope name. A root scope name has to match the following pattern: ^[A-Za-z0-9.][A-Za-z0-9_.\\/>-]*$`

You can find my code here: [GitHub Compare](https://github.com/ultralytics/ultralytics/compare/main...adamp87:ultralytics:tflite_accurate)

Thanks for your help

**Comment by EpiX-1** [comment]
I'm able to reproduce the issue. Your code seems good to me. 
I've managed to resolve the issue by reverting to an [older commit of YOLOv8](https://github.com/ultralytics/ultralytics/tree/b507e3a03260bc333474641b7d36aaabb736b8af). I have no idea why it occurs, though.

**Comment by adamp87** [comment]
Thanks for your reply. If you revert to an older commit you remove the code
with the QuantizationDebugger

**Comment by adamp87** [comment]
Sorry my bad, thanks for the clarification. I will try it the next days and
try to find what caused the problem. Lets keep in touch

**Comment by adamp87** [comment]
Hi @EpiX-1,

I modified my branch not to export the ONNX model. First I export the ONNX model using the branch you mentioned then I execute my code and I still get the error message. Can you maybe post your code? Or it might be different version of the converter tools?

Here is how I executed in a Colab:
```
!git clone https://github.com/adamp87/ultralytics.git
%cd /content/ultralytics
!pip install -e .
!git checkout b507
!yolo export model=yolov8n.pt format=onnx
!git checkout tflite_accurate
!yolo export model=yolov8n.pt data=coco128.yaml format=tflite imgsz=640
```

Thanks!

**Comment by EpiX-1** [comment]
Hi @adamp87 , 

I was wrong regarding the origin of the problem. It has nothing to do with the version of YOLOv8 and it's more likely related to the environment used. I sincerely apologize for this mistake. 

My previous workflow was the following:  
- Convert the.pt model to .onnx thanks to YOLOv8 export code.  
- Convert manually the .onnx model to keras thanks to onnx2tf (docker)  
- Convert and do selective quantization on the keras model thanks to a custom script using the QuantizationDebugger. This script was executed in another conda environment than the one used to export the .pt model to .onnx. Probably that's why I didn't face the issue before. 

I've included a part of the selective quantization script I use in YOLOv8 source code. You can find it [here](https://github.com/EpiX-1/ultralytics/commit/fb18e3e224d286aa27200890a7894f011ea0b33d#). You can also look at this [colab script](https://github.com/PINTO0309/onnx2tf/files/14531514/selective_quantization.zip) if you want to export your model.

**Comment by adamp87** [comment]
Hi @EpiX-1,

the conversion with excluded operations is working, thank you!

First, I tried downloading the newest versions of the necessary tools to convert from onnx to Keras(h5). This gave me an error during the conversion, "problem with Resize". Then I used the Docker version to convert from onnx to Keras(h5), this worked. So I believe this must be some version issue, but with Docker it is working. After a few modifications in my code I was able to have a custom model. Still, I have the feeling it is quite cumbersome to give all the ops from the YOLOv8 Detection. I also couldn't manage to remove the Quantize after the Logistic, after a Quantize it does immediately a Dequantize.

I would like to thank you big time for helping me out and guiding me the right way. Thank you so much!
